File: src/modules/execution-control/services/execution-abort-service.ts
// src/modules/execution-control/services/execution-abort-service.ts
// module: execution-control | layer: services | role: 执行中止服务
// summary: 提供脚本执行中止功能，包含前端状态管理和后端通信

import { invoke } from '@tauri-apps/api/core';
import { message } from 'antd';

export interface ExecutionAbortRequest {
  reason?: string;
  force?: boolean;
}

export interface ExecutionAbortResult {
  success: boolean;
  message: string;
  stoppedAt?: {
    stepIndex: number;
    stepName: string;
  };
}

/**
 * 执行中止服务
 * 负责协调前端状态和后端中止操作
 */
export class ExecutionAbortService {
  private static instance: ExecutionAbortService;
  private abortController: AbortController | null = null;
  private currentExecutionId: string | null = null;

  static getInstance(): ExecutionAbortService {
    if (!ExecutionAbortService.instance) {
      ExecutionAbortService.instance = new ExecutionAbortService();
    }
    return ExecutionAbortService.instance;
  }

  /**
   * 开始执行时注册执行ID
   */
  startExecution(executionId: string): void {
    this.currentExecutionId = executionId;
    this.abortController = new AbortController();
    console.log(`🎬 [执行控制] 注册执行ID: ${executionId}`);
  }

  /**
   * 执行完成时清理
   */
  finishExecution(): void {
    console.log(`🏁 [执行控制] 清理执行状态: ${this.currentExecutionId}`);
    this.currentExecutionId = null;
    this.abortController = null;
  }

  /**
   * 检查是否有活跃的执行
   */
  hasActiveExecution(): boolean {
    return this.currentExecutionId !== null;
  }

  /**
   * 获取当前执行ID
   */
  getCurrentExecutionId(): string | null {
    return this.currentExecutionId;
  }

  /**
   * 中止当前执行
   */
  async abortExecution(request: ExecutionAbortRequest = {}): Promise<ExecutionAbortResult> {
    if (!this.hasActiveExecution()) {
      const result: ExecutionAbortResult = {
        success: false,
        message: '没有正在执行的脚本'
      };
      return result;
    }

    const executionId = this.currentExecutionId!;
    const reason = request.reason || '用户手动中止';
    
    console.log(`🛑 [执行控制] 开始中止执行: ${executionId}, 原因: ${reason}`);

    try {
      // 1. 设置前端中止信号
      if (this.abortController) {
        this.abortController.abort(reason);
        console.log(`🚫 [执行控制] 前端中止信号已发送`);
      }

      // 2. 调用后端中止接口
      const backendResult = await this.abortBackendExecution(executionId, request);
      
      // 3. 清理状态
      this.finishExecution();

      const result: ExecutionAbortResult = {
        success: true,
        message: `执行已中止: ${reason}`,
        stoppedAt: backendResult.stoppedAt
      };

      console.log(`✅ [执行控制] 中止完成:`, result);
      message.warning(`🛑 ${result.message}`, 5);
      
      return result;

    } catch (error) {
      console.error(`❌ [执行控制] 中止失败:`, error);
      
      // 即使后端调用失败，也要清理前端状态
      this.finishExecution();

      const errorMessage = error instanceof Error ? error.message : '未知错误';
      const result: ExecutionAbortResult = {
        success: false,
        message: `中止失败: ${errorMessage}`
      };

      message.error(`❌ ${result.message}`, 5);
      return result;
    }
  }

  /**
   * 调用后端中止接口
   */
  private async abortBackendExecution(
    executionId: string, 
    request: ExecutionAbortRequest
  ): Promise<{ stoppedAt?: { stepIndex: number; stepName: string } }> {
    try {
      // 尝试调用后端中止接口
      const result = await invoke('abort_script_execution', {
        executionId,
        reason: request.reason || '用户手动中止',
        force: request.force || false
      });

      console.log(`🎯 [执行控制] 后端中止成功:`, result);
      return result as any;

    } catch (error) {
      console.warn(`⚠️ [执行控制] 后端中止接口调用失败，可能后端不支持该功能:`, error);
      
      // 如果后端不支持中止接口，尝试其他方式
      try {
        // 尝试调用通用中止接口
        await invoke('cancel_current_operation');
        console.log(`🔄 [执行控制] 使用通用中止接口成功`);
        return {};
      } catch (secondError) {
        console.warn(`⚠️ [执行控制] 通用中止接口也失败:`, secondError);
        
        // 最后尝试：强制终止所有 ADB 操作
        try {
          await invoke('force_stop_all_adb_operations');
          console.log(`🔨 [执行控制] 强制停止 ADB 操作成功`);
          return {};
        } catch (thirdError) {
          console.error(`❌ [执行控制] 所有中止方式都失败:`, thirdError);
          throw new Error('无法中止后端执行，请手动重启应用');
        }
      }
    }
  }

  /**
   * 获取中止信号（用于执行过程中检查）
   */
  getAbortSignal(): AbortSignal | null {
    return this.abortController?.signal || null;
  }

  /**
   * 检查是否已被中止
   */
  isAborted(): boolean {
    return this.abortController?.signal.aborted || false;
  }
}

File: src-tauri/src/services/execution_abort_service.rs
#![allow(dead_code)]
#![allow(unused_variables)]
#![allow(unused_imports)]

// src-tauri/src/services/execution_abort_service.rs
// module: services | layer: services | role: 后端执行中止服务
// summary: 提供真正的后端 ADB 操作中止功能

use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::process::Command;
use serde::{Deserialize, Serialize};
use tauri::command;
use tracing::{info, warn, error};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AbortRequest {
    pub execution_id: String,
    pub reason: Option<String>,
    pub force: Option<bool>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AbortResult {
    pub success: bool,
    pub message: String,
    pub stopped_at: Option<StepLocation>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StepLocation {
    pub step_index: usize,
    pub step_name: String,
}

// 全局执行状态管理器
lazy_static::lazy_static! {
    static ref EXECUTION_MANAGER: Arc<Mutex<ExecutionManager>> = 
        Arc::new(Mutex::new(ExecutionManager::new()));
}

#[derive(Debug)]
pub struct ExecutionManager {
    // 当前活跃的执行ID
    active_executions: HashMap<String, ExecutionContext>,
}

#[derive(Debug)]
pub struct ExecutionContext {
    pub execution_id: String,
    pub device_id: String,
    pub started_at: std::time::Instant,
    pub current_step: Option<StepLocation>,
    pub should_abort: bool,
}

impl ExecutionManager {
    pub fn new() -> Self {
        Self {
            active_executions: HashMap::new(),
        }
    }

    pub fn register_execution(&mut self, execution_id: String, device_id: String) {
        info!("🎬 [执行管理器] 注册执行: {} (设备: {})", execution_id, device_id);
        
        let context = ExecutionContext {
            execution_id: execution_id.clone(),
            device_id,
            started_at: std::time::Instant::now(),
            current_step: None,
            should_abort: false,
        };

        self.active_executions.insert(execution_id, context);
    }

    pub fn update_current_step(&mut self, execution_id: &str, step: StepLocation) {
        if let Some(context) = self.active_executions.get_mut(execution_id) {
            context.current_step = Some(step);
        }
    }

    pub fn mark_for_abort(&mut self, execution_id: &str) -> bool {
        if let Some(context) = self.active_executions.get_mut(execution_id) {
            context.should_abort = true;
            info!("🛑 [执行管理器] 标记执行中止: {}", execution_id);
            true
        } else {
            warn!("⚠️ [执行管理器] 未找到执行: {}", execution_id);
            false
        }
    }

    pub fn should_abort(&self, execution_id: &str) -> bool {
        self.active_executions
            .get(execution_id)
            .map(|ctx| ctx.should_abort)
            .unwrap_or(false)
    }

    pub fn finish_execution(&mut self, execution_id: &str) -> Option<ExecutionContext> {
        info!("🏁 [执行管理器] 完成执行: {}", execution_id);
        self.active_executions.remove(execution_id)
    }

    pub fn list_active_executions(&self) -> Vec<&ExecutionContext> {
        self.active_executions.values().collect()
    }
}

/// 中止脚本执行
#[command]
pub async fn abort_script_execution(request: AbortRequest) -> Result<AbortResult, String> {
    info!("🛑 [中止服务] 收到中止请求: {:?}", request);

    let execution_id = &request.execution_id;
    let reason = request.reason.unwrap_or_else(|| "用户手动中止".to_string());
    let force = request.force.unwrap_or(false);

    // 1. 标记执行应当中止
    let (device_id, current_step) = {
        let mut manager = EXECUTION_MANAGER.lock().map_err(|e| {
            error!("❌ [中止服务] 获取执行管理器锁失败: {}", e);
            "无法获取执行管理器锁".to_string()
        })?;

        if !manager.mark_for_abort(execution_id) {
            return Ok(AbortResult {
                success: false,
                message: "没有找到指定的执行".to_string(),
                stopped_at: None,
            });
        }

        let context = manager.active_executions.get(execution_id);
        let device_id = context.as_ref().map(|c| c.device_id.clone()).unwrap_or_default();
        let current_step = context.as_ref().and_then(|c| c.current_step.clone());

        (device_id, current_step)
    };

    // 2. 尝试中止设备上的 ADB 操作
    let adb_abort_success = if force {
        force_kill_adb_operations(&device_id).await
    } else {
        graceful_abort_adb_operations(&device_id).await
    };

    if !adb_abort_success {
        warn!("⚠️ [中止服务] ADB 操作中止失败，但前端状态已标记中止");
    }

    // 3. 清理执行状态
    {
        let mut manager = EXECUTION_MANAGER.lock().map_err(|e| {
            error!("❌ [中止服务] 清理时获取执行管理器锁失败: {}", e);
            "无法清理执行状态".to_string()
        })?;
        manager.finish_execution(execution_id);
    }

    Ok(AbortResult {
        success: true,
        message: format!("执行已中止: {}", reason),
        stopped_at: current_step,
    })
}

/// 优雅中止 ADB 操作
async fn graceful_abort_adb_operations(device_id: &str) -> bool {
    info!("🔄 [ADB中止] 尝试优雅中止设备 {} 的操作", device_id);

    // 尝试发送中断信号给当前正在运行的 ADB 进程
    match Command::new("adb")
        .args(["-s", device_id, "shell", "input", "keyevent", "KEYCODE_BACK"])
        .output()
    {
        Ok(output) => {
            if output.status.success() {
                info!("✅ [ADB中止] 优雅中止成功 (发送返回键)");
                true
            } else {
                warn!("⚠️ [ADB中止] 优雅中止失败: {}", String::from_utf8_lossy(&output.stderr));
                false
            }
        }
        Err(e) => {
            error!("❌ [ADB中止] 优雅中止命令执行失败: {}", e);
            false
        }
    }
}

/// 强制终止 ADB 操作
async fn force_kill_adb_operations(device_id: &str) -> bool {
    info!("🔨 [ADB中止] 强制终止设备 {} 的操作", device_id);

    // 方法1: 终止设备上当前的操作
    let kill_current = Command::new("adb")
        .args(["-s", device_id, "shell", "pkill", "-f", "input"])
        .output();

    // 方法2: 重启 ADB 连接
    let restart_adb = Command::new("adb")
        .args(["disconnect", device_id])
        .output()
        .and_then(|_| {
            std::thread::sleep(std::time::Duration::from_millis(500));
            Command::new("adb")
                .args(["connect", device_id])
                .output()
        });

    match (kill_current, restart_adb) {
        (Ok(_), Ok(_)) => {
            info!("✅ [ADB中止] 强制终止成功");
            true
        }
        _ => {
            error!("❌ [ADB中止] 强制终止失败");
            false
        }
    }
}

/// 取消当前操作 (通用接口)
#[command]
pub async fn cancel_current_operation() -> Result<String, String> {
    info!("🔄 [通用中止] 取消当前操作");

    let active_executions = {
        let manager = EXECUTION_MANAGER.lock().map_err(|e| {
            error!("❌ [通用中止] 获取执行管理器锁失败: {}", e);
            "无法获取执行管理器锁".to_string()
        })?;
        manager.list_active_executions().iter().map(|ctx| ctx.execution_id.clone()).collect::<Vec<_>>()
    };

    if active_executions.is_empty() {
        return Ok("没有活跃的执行需要取消".to_string());
    }

    let mut results = Vec::new();
    for execution_id in active_executions {
        let request = AbortRequest {
            execution_id: execution_id.clone(),
            reason: Some("通用取消操作".to_string()),
            force: Some(false),
        };

        match abort_script_execution(request).await {
            Ok(_) => results.push(format!("✅ 取消执行: {}", execution_id)),
            Err(e) => results.push(format!("❌ 取消失败: {} - {}", execution_id, e)),
        }
    }

    Ok(results.join("; "))
}

/// 强制停止所有 ADB 操作
#[command]
pub async fn force_stop_all_adb_operations() -> Result<String, String> {
    info!("🔨 [强制停止] 停止所有 ADB 操作");

    // 1. 标记所有执行中止
    let device_ids = {
        let mut manager = EXECUTION_MANAGER.lock().map_err(|e| {
            error!("❌ [强制停止] 获取执行管理器锁失败: {}", e);
            "无法获取执行管理器锁".to_string()
        })?;

        let device_ids: Vec<String> = manager.active_executions
            .values()
            .map(|ctx| ctx.device_id.clone())
            .collect();

        // 标记所有执行中止
        for execution_id in manager.active_executions.keys().cloned().collect::<Vec<_>>() {
            manager.mark_for_abort(&execution_id);
        }

        device_ids
    };

    // 2. 强制终止所有设备的 ADB 操作
    let mut results = Vec::new();
    for device_id in device_ids {
        if force_kill_adb_operations(&device_id).await {
            results.push(format!("✅ 强制停止设备: {}", device_id));
        } else {
            results.push(format!("❌ 停止失败设备: {}", device_id));
        }
    }

    // 3. 清理所有执行状态
    {
        let mut manager = EXECUTION_MANAGER.lock().map_err(|e| {
            error!("❌ [强制停止] 清理时获取执行管理器锁失败: {}", e);
            "无法清理执行状态".to_string()
        })?;
        manager.active_executions.clear();
    }

    if results.is_empty() {
        Ok("没有需要停止的操作".to_string())
    } else {
        Ok(results.join("; "))
    }
}

/// 检查执行是否应当中止 (供其他模块调用)
pub fn should_abort_execution(execution_id: &str) -> bool {
    EXECUTION_MANAGER
        .lock()
        .map(|manager| manager.should_abort(execution_id))
        .unwrap_or(false)
}

/// 注册新的执行 (供其他模块调用)
pub fn register_execution(execution_id: String, device_id: String) {
    if let Ok(mut manager) = EXECUTION_MANAGER.lock() {
        manager.register_execution(execution_id, device_id);
    }
}

/// 完成执行 (供其他模块调用)
pub fn finish_execution(execution_id: &str) {
    if let Ok(mut manager) = EXECUTION_MANAGER.lock() {
        manager.finish_execution(execution_id);
    }
}